src/python/pants/core/util_rules/wrap_source.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+# This file should be moved once we figure out where everything belongs
+
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+from typing import Iterable, Union
+
```

Comment:
From a UX perspective, it feels like this should draw inspiration from `http_source` instead: https://www.pantsbuild.org/v2.15/docs/reference-resource#codesourcecode ... such that the syntax was more like:
```python
experimental_shell_command(
  name='cmd',
)

python_sources(
  sources=output_sources(":cmd"),
)
```

I know that @thejcannon also took advantage of codegen for the purposes of `http_source`, but if we need to deepen the support for alternative sources fields types (because due to its implementation, `http_source` is currently limited to `resource` and `file`) that seems like it would still be worth doing.

Reply:
Ah yeah, see I knew roping Stu in was the right call. I like this syntax as well (bikeshed on the name).

Reply:
So I've been mulling this over for a while, and having looked at `http_source`, I can see how it works, and could apply here. I'm happy to implement it if the discussion lands on implementing it in this way.

I've re-written this comment a few times, to make sure that my understanding of how things work lines up with what's actually going onside Pants. Some of the objections I initially had turned out to not be entirely correct. 

I think there's some complication however, that justifies the conceptual separation of our `sources` target generators and a corresponding multiple source wrapper targets. Count me as a medium _no_ against adding the new value type.


Most `x_sources` file types (which produce `x_source` targets) are `TargetFilesGenerators`, which currently produce `GeneratedTargets` representing one target per file matched by the generator. Eventually, the `MultipleSourcesField` on the target is used in the `GeneratedTargets`-producing rule to resolve a `SourcesPaths`, which does not consider codegen.

It would be possible to amend that `-> GeneratedTargets` rule to act as a no-op if the `MultipleSourcesField`'s value were an `output_sources`. I presume it would be possible to implement a `GeneratedSources` rule for each of the `MultipleSourcesField`s that we end up wanting to support. Rules would need to be registered per `x_sources` target type, since codegen requires a single rule per `(InputSourceField, OutputSourceField)` pair. This can't be as neat a solution as the `TargetFilesGenerator` approach is for generating targets for actual files. 

The conceptual gap here is the `x_sources` Target would be a Target Generator when operating on source files, but a plain old `Target` that resolves to several (codegenned) source files when using an `output_sources` value. This makes it difficult to reason about the behaviour of this target type, and similarly difficult to document: This would significantly complicate the documentation for each of these `x_sources` targets. It would need to _explain_ that the target behaves significantly differently depending on whether you're using it for real files or to wrap generated files from elsewhere.

It also looks as though the behaviour of `moved_fields` and `copied_fields` would also be somewhat difficult to reason about -- they'd be neither moved, nor copied if the target is used directly? It seems like `parametrize` also behaves weirdly here.

So I think an implementation using the existing target generators with a new `output_sources` value type would (at the moment) comfortably land in the "difficult to explain" bucket. We are doing something significantly different to what the target types currently do.

On the other hand, the current implementation produces a bunch of target types (not ideal), but they're easier to reason about, and they're all grouped in help under `experimental_wrap_as`, which should make them easy-ish to spot in the help. 


**This is `experimental`, we can change the API later**

Currently all of the wrapper types will be marked `experimental`, which would allow us time to deal with some of the above items where possible. With that in mind, we could land this as currently proposed, and take the time to solve the above technical blockers.

Reply:
Specifically I would assume this acts more like a single python_source.
Perhaps we can encode that requirement? 1 source and one target.

Reply:
@thejcannon I don't think that's a good restriction to make here -- you could have an `esc` that generates a pile of files for use as `resources`, and it may not be possible to enumerate them all. Alternatively, you could have a process that generates a number of source files (not unreasonable in JVM land), and you need to include all of them. If you had 1 source per target, you'd end up with a lot of individual target definitions, and you'd need to manually map the dependencies between them.

Reply:
Hmm I think I see the core of this now. I think the issue here is we're trying to model something completely new and we just don't have the machinery to express it.

Ideally this _doesnt_ require any new targets and we say that this python source comes from this thing.

I think the python issue is interesting because we could have one-target-many-files, but the only target for that today is a target generator, so no dice. The only other target is one-file-only. So also no dice. 

Maybe the easiest path forward is relaxing the latter and allowing python_source to be multiple files? That's technically what you have here. A generator generating a PythonSourceField (singular)

Reply:
I read the above as having a new kind of source field type, that acts like a hybrid between a regular source field and the dependencies field.

```python
experimental_shell_command(
  name='cmd',
  ..,
)

python_sources(
  generated_sources=[":cmd"],
)
```

>  I think the issue here is we're trying to model something completely new and we just don't have the machinery to express it.

I mostly agree. We do have something similar since before, but it's not ergonomic to use. For the Docker build context for instance, we pull in `files` sources from the `docker_image` dependencies, and any `esc` targets may be used thanks to codegen.

I have a nagging feeling that we may want to look at our codegen API however--that it requires a source field is a nuance for one, and over and over again we're jumping through hoops to accommodate its API when being applied in new and interesting ways (but that's for another issue).

Reply:
> The conceptual gap here is the `x_sources` Target would be a Target Generator when operating on source files, but a plain old `Target` that resolves to several (codegenned) source files when using an `output_sources` value. This makes it difficult to reason about the behaviour of this target type, and similarly difficult to document: This would significantly complicate the documentation for each of these `x_sources` targets. It would need to _explain_ that the target behaves significantly differently depending on whether you're using it for real files or to wrap generated files from elsewhere.

`TargetGenerator` is an API that we provide to target implementors, rather than necessarily limiting what we can do with targets. New logic can be added in [the spot where we implement `TargetGenerators`](https://github.com/pantsbuild/pants/blob/dfbfd75f3b1ce1d4c53a639eaeda2254d75078d8/src/python/pants/engine/internals/graph.py#L212-L214) to trigger (*a form of: see below) codegen.

I'm not sure that there would be a mode switch in this case (and so no extra documentation for this case): if multiple files were provided, the target could act exactly as it would otherwise, by generating multiple `python_source` targets.

* The primary complication I see is that this should likely not use the codegen API, per-se. Instead, it seems like a separate `@union` for "things which can be used as source file inputs" could be added and used both for this use case, and for `http_source`. We don't necessarily want all codegen to be able to during the calculation of dependencies, and that would keep the graph simpler.

Reply:
> I think the issue here is we're trying to model something completely new and we just don't have the machinery to express it.

@thejcannon We do have the machinery to express it; this PR demonstrates it. We don't have great machinery to do it as a source input _yet_.

> Maybe the easiest path forward is relaxing the latter and allowing `python_source` to be multiple files?

@thejcannon I think this would make the difference between `python_source` and `python_sources` conceptually very difficult to understand -- my hunch is that newcomers would accidentally start using multi-file `python_source` targets more often and lose the benefits of fine-grained dependencies and caching. I'm very keen that this change _not_ make the core concepts of pants less easy to understand.

> I'm not sure that there would be a mode switch in this case (and so no extra documentation for this case): if multiple files were provided, the target could act exactly as it would otherwise, by generating multiple python_source targets.

@stuhood Unless I'm missing something, this would require the user to enumerate the sources from the input file, and would not allow the user to supply a glob, because the files matched by a glob wouldn't be known until after the input target has produced its outputs (potentially after the process has run).

Reply:
> @stuhood Unless I'm missing something, this would require the user to enumerate the sources from the input file, and would not allow the user to supply a glob, because the files matched by a glob wouldn't be known until after the input target has produced its outputs (potentially after the process has run).

Have thought about this some more, and you're right that "knowing the list of files that will be generated" is an important distinction in this case... but not the only one. The actual content of the files is also important for dependency inference -- but due to cost concerns (to infer dependencies for most languages, you'd have to generate all code in the repository), codegen doesn't currently run as part of dependency inference.

----

Currently, code generation (`GenerateSourcesRequest`):
* Does not need to know ahead of time which files it will generate
* Does not run before/as-part-of `Dependencies` calculations (because dependency inference implementations avoid calling it)
    * Instead, if you want dependencies inferred for codegen'd targets, you need to add dependency inference for the `gen+lang` pair: e.g. `thrift+python`.
    * That means that, as written, this PR will not have dependency inference applied. 

Meanwhile, target generation (`TargetGenerator`, and in particular `TargetFilesGenerator`):
* Does need to ["know" ahead of time](https://github.com/pantsbuild/pants/blob/851998457bb87d2397c9f041fa372fbbd97f2473/src/python/pants/engine/internals/graph.py#L1420-L1447) which files/targets it will generate
    * The "know" here mainly refers to the fact that `TargetFilesGenerator` uses a relatively light operation to list files (`SourcesPaths`) rather than snapshotting them.
* Does run before `Dependencies` calculations
* Participates in dependency inference by actually capturing the files.
    * In the case of `TargetFilesGenerator` this is snapshotting files off of disk.

----

I think that users generating code using `experimental_shell_command` (or via `http_source`) would expect dependency inference to be applied (unlike with "native" `GenerateSourcesRequest` implementations, where there is an expectation that the `@rule` author implements inference in a generation specific way). And as mentioned above, I think that generic codegen is probably too expensive to run as part of dependencies calculations.

So to refine my suggestion: hooking this in to `TargetFilesGenerator`, using either codegen (if it doesn't make the graph too complex: unclear), or a new more limited `@union` seems like a path forward.

Reply:
Thanks for confirming my technical understanding of `GenerateSourcesRequest` and `TargetFilesGenerator`.

Per my previous comment here, I think that needing to know what files will be generated ahead of time is of limited utility and creates a maintenance burden.

Currently, an `x_sources` target generator needs to be written once, and will keep up to date as more `x` files are added to a directory. Using a `TargetFilesGenerator` that subsets the outputs from another target would mean the `BUILD` file would need to be updated whenever the contents of the target would change. Admittedly, this would guarantee the best dependency/cache performance, however it would come at the expense of a maintenance burden of the build file, and the build errors that arise due to files not being captured from the outputs would be hard to debug.

On the other hand, the implementation in the PR can glob the outputs of another target, so for cases where _all_ of the outputs are needed, the `BUILD` file only needs to be written once, and remains straightforward. If you would benefit from more fine-grained targets, then you can can add new targets for each subset of files. The user has the option of choosing easy maintenance or fine-grained caching.

At the moment, for most of the use cases I'm imagining

In my view, dependency inference is a nice-to-have for this use case, but not essential: in the `experimental_shell_command`/`experimental_run_in_sandbox` use cases, users have to specify dependencies manually.

As a firmer technical red flag here, for `esc` use cases: Java's dependency inference code in particular would need to run the underlying `esc` process to fulfil [the `SourceFilesRequest`s in its first-party symbol mapping](https://github.com/pantsbuild/pants/blob/main/src/python/pants/backend/java/dependency_inference/symbol_mapper.py#L45-L47). That would mean _every_ `esc` that fulfils a `_sources` target would need to run before dependency inference can be performed, regardless of whether that dependency is necessary to fulfil the goal.

My preference here is for this idea to be wrapped up in an `experimental` API with minimal surface area, so that we can get a better idea about how this is being used, and then proceed with a deeper change in machinery once we've got a more solid understanding of how it'll be used.

Reply:
> Currently, an `x_sources` target generator needs to be written once, and will keep up to date as more `x` files are added to a directory. Using a `TargetFilesGenerator` that subsets the outputs from another target would mean the `BUILD` file would need to be updated whenever the contents of the target would change. Admittedly, this would guarantee the best dependency/cache performance, however it would come at the expense of a maintenance burden of the build file, and the build errors that arise due to files not being captured from the outputs would be hard to debug.

I am not proposing that this integration would require explicitly listing the files to generate targets for, as I don't think that that buys you a significant difference in performance. Basically: if you required that all files be explicitly listed, `./pants list` would be able to skip generating code to generate targets, but `./pants dependencies` would need to generate code in order to compute dependencies for inference.

> In my view, dependency inference is a nice-to-have for this use case, but not essential: in the `experimental_shell_command`/`experimental_run_in_sandbox` use cases, users have to specify dependencies manually.

To be clear: this is about both the dependencies of the `wrapped_as_` target, and dependencies _on_ the `wrapped_as_` target.

-----

I might be fine with landing the wrappers as experimental, as long as we agree that they are even less stable than `experimental_shell_command` itself, and should probably stabilize at a different time.

Reply:
> but `./pants dependencies` would need to generate code in order to compute dependencies for inference

Yup, that's the red flag I had above, you would need to run the processes in order to compute dependencies, and _that_ would have an impact on performance: all source-providing targets would need to be evaluated before the dependency inference would take place. This could have a significant impact on performance.

Reply:
> I might be fine with landing the wrappers as experimental, as long as we agree that they are even less stable than `experimental_shell_command itself`, and should probably stabilize at a different time.

Yes, I agree entirely.

To be clear, I think there's value in supporting alternative values for `SourceField`s, but at the moment, but it's probably a bigger design task with riskier scope than we can solve in a PR discussion.

Reply:
> > but `./pants dependencies` would need to generate code in order to compute dependencies for inference
> 
> Yup, that's the red flag I had above, you would need to run the processes in order to compute dependencies, and _that_ would have an impact on performance: all source-providing targets would need to be evaluated before the dependency inference would take place. This could have a significant impact on performance.

Only those actually used in this position: not all codegen targets. But yea.

I could imagine this usecase as differentiated from how we handle "native" codegen: native codegen has a higher implementation difficulty bar, because you need to implement both dependency inference rules and the codegen itself.

> To be clear, I think there's value in supporting alternative values for `SourceField`s, but at the moment, but it's probably a bigger design task with riskier scope than we can solve in a PR discussion.

I disagree that a PR is necessarily the wrong place to have this discussion. To _reduce_ the need for this type of discussion on PRs it can be useful to ensure that design/UX are discussed before a PR is started... but in many cases, some of that ends up happening during code review.